### ISSUE TITLE
Offer alternative EnumConverter constructor accepting Function<U, T> as a Converter::to implementation

### ISSUE THREAD

**Issue by juriad** [jOOQ/jOOQ#12423]
### Use case:

We have quite a few enums that look like this. The enum constants contain abbreviation which should be used for storing in DB). We don't use code generation because of dynamic schema.
```java
@Getter
@RequiredArgsConstructor
public enum NodeKind {
    NEW_DRAFT("dn"),
    CHANGE_DRAFT("dc"),
    DELETE_DRAFT("dd"),

    PUBLISHED_WITH_CHANGE_DRAFT("pc"),
    PUBLISHED_WITH_DELETE_DRAFT("pd"),
    PUBLISHED_WITHOUT_DRAFT("p"),

    HISTORIC_DELETED("hd"),
    HISTORIC("h");

    private final String abbreviation;
}
```

`org.jooq.impl.EnumConverter` is meant to be extended; it specifically mentions:
> Subclasses may override this method to provide a custom reverse mapping implementation

This extension is however impossible if it should rely on any extra constructor parameter.
We wanted to define `GeneralizedEnumConverter<T, U extends Enum<U>>` which would apply `Function<U, ?> keyExtractor` to each enum constant.
Because the map `lookup` is initialized in  the constructor, any subclass cannot alter it. It can only change implementation of `to` but there it has limited options; it cannot touch any field (either it is `private` in `EnumConverter` or it has not been initialized yet).

### Possible solution you'd like to see:

I would like `EnumConverter` to have built-in support for `keyExtractor` as the most general mapping principle.
This change should be backwards compatible.

```java
    private final Map<T, U> lookup;
    private final Function<U, ?> keyExtractor; // no need for Enum ORDINAL, STRING

    public GeneralizedEnumConvertor(Class<T> fromType, Class<U> toType) {
        this(fromType, toType,
                // [#8045] Also support Kotlin Int type (which translates to int.class)
                Number.class.isAssignableFrom(wrapper(fromType)) ? Enum::ordinal : Enum::name
        );
    }

    public GeneralizedEnumConvertor(Class<T> fromType, Class<U> toType, Function<U, ?> keyExtractor) {
        super(fromType, toType);

        this.keyExtractor = keyExtractor;
        this.lookup = new LinkedHashMap<>();

        for (U u : toType.getEnumConstants())
            this.lookup.put(to(u), u);
    }

...

    @Override
    public T to(U userObject) {
        if (userObject == null)
            return null;
        return convert(keyExtractor.apply(userObject), fromType());
    }
```

Using it is easy:
```java
    public static final DataType<NodeKind> KIND_DATA_TYPE = SQLDataType.VARCHAR(2).nullable(false).asConvertedDataType(
            new GeneralizedEnumConvertor<>(String.class, NodeKind.class, NodeKind::getAbbreviation));
```

### Possible workarounds:

Every user must copy-paste `EnumConverter` and implement the change in their code. Alternatively they can start with `AbstractConverter` and end with a very similar solution.

### Versions:

- jOOQ: 3.15.2
- Java: 11
- Database (include vendor): Postgres
- OS: Linux
- JDBC Driver (include name if inofficial driver):

**Comment by lukaseder** [comment]
Thanks a lot for your suggestion. Yes, I agree we could do this

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.16.0
